File: mathewDavidov/assignments/treesBFS/lc107/lc107.cpp
// Problem Statement #

// Given a binary tree, populate an array to represent its level-by-level traversal in reverse order, i.e., the lowest level comes first. You should populate the values of all nodes in each level from left to right in separate sub-arrays.

using namespace std;

#include <deque>
#include <iostream>
#include <queue>

class TreeNode {
 public:
  int val = 0;
  TreeNode *left;
  TreeNode *right;

  TreeNode(int x) {
    val = x;
    left = right = nullptr;
  }
};

class ReverseLevelOrderTraversal {
 public:
  static deque<vector<int>> traverse(TreeNode *root) {
    deque<vector<int>> result = deque<vector<int>>();
    // TODO: Write your code here
    return result;
  }
};
int main(int argc, char *argv[]) {
  TreeNode *root = new TreeNode(12);
  root->left = new TreeNode(7);
  root->right = new TreeNode(1);
  root->left->left = new TreeNode(9);
  root->right->left = new TreeNode(10);
  root->right->right = new TreeNode(5);
  auto result = ReverseLevelOrderTraversal::traverse(root);

  cout << "Reverse level order traversal: ";
  for (auto que : result) {
    for (auto num : que) {
      cout << num << " ";
    }
    cout << endl;
  }
}



// Solution
// -----
// class ReverseLevelOrderTraversal {
//  public:
//   static deque<vector<int>> traverse(TreeNode *root) {
//     deque<vector<int>> result = deque<vector<int>>();
//     if (root == nullptr) {
//       return result;
//     }

//     queue<TreeNode *> queue;
//     queue.push(root);
//     while (!queue.empty()) {
//       int levelSize = queue.size();
//       vector<int> currentLevel;
//       for (int i = 0; i < levelSize; i++) {
//         TreeNode *currentNode = queue.front();
//         queue.pop();
//         // add the node to the current level
//         currentLevel.push_back(currentNode->val);
//         // insert the children of current node to the queue
//         if (currentNode->left != nullptr) {
//           queue.push(currentNode->left);
//         }
//         if (currentNode->right != nullptr) {
//           queue.push(currentNode->right);
//         }
//       }
//       // append the current level at the beginning
//       result.push_front(currentLevel);
//     }

//     return result;
//   }
// };

// -----

// Time complexity #
// The time complexity of the above algorithm is O(N), where ‘N’ is the total number of nodes in the tree. This is due to the fact that we traverse each node once.

// Space complexity #
// The space complexity of the above algorithm will be O(N) as we need to return a list containing the level order traversal. We will also need O(N) space for the queue. Since we can have a maximum of N/2 nodes at any level (this could happen only at the lowest level), therefore we will need O(N) space to store them in the queue.

File: mathewDavidov/assignments/treesBFS/lc107/lc107.java
/**
 * Problem Statement:
 * Given a binary tree, populate an array to represent its level-by-level traversal in reverse order,
 * i.e., the lowest level comes first. 
 * You should populate the values of all nodes in each level from left to right in separate sub-arrays.
 * 
 * Example:
 *      3
 *     9 20
 *      15 7
 * -> [[15, 7], [9, 20], [3]]
 * 
 * Edge cases:
 * null root -> empty list
 * 
 * As the problem is similar to normal level-by-level traversal,
 * to get the values from leaf to root, we can append our arrays 
 * at index 0 every time we finish a level.
 * 
 * initialize the list containing the level-by-level traversal (use linkedlist, inserting on the ends is O(1))
 * initialize a queue and add the root
 * while the queue is not empty:
 *   retrieve the current size of the queue in a variable
 *   initialize a list for the current level
 *   loop from 0 to the size of the queue (in the above variable):
 *     poll the element in the front in the queue
 *     if its left child exists, add it to the queue
 *     if its right child exists, add it to the queue
 *     add the current node's value (that was polled) to the current level list
 *   add the current level list to the front of the level-by-level list
 * return the level-by-level list
 * 
 * Runtime:
 * Time: O(n)
 * Space: O(n)
 */
import java.util.*;

public class lc107 {
    public static List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> levelByLevelList = new LinkedList<>();

        if (root == null) {
            return levelByLevelList;
        }

        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int queueSize = queue.size();
            List<Integer> levelList = new ArrayList<>();

            for (int i = 0; i < queueSize; i++) {
                TreeNode current = queue.poll();

                if (current.left != null) {
                    queue.offer(current.left);
                }

                if (current.right != null) {
                    queue.offer(current.right);
                }

                levelList.add(current.val);
            }

            levelByLevelList.add(0, levelList);
        }

        return levelByLevelList;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(12);
        root.left = new TreeNode(7);
        root.right = new TreeNode(1);
        root.left.left = new TreeNode(9);
        root.right.left = new TreeNode(10);
        root.right.right = new TreeNode(5);
        List<List<Integer>> result = levelOrderBottom(root);
      
        System.out.println("Reverse level order traversal: ");
        System.out.println(result);
    }
}

class TreeNode {
    public int val = 0;
    public TreeNode left;
    public TreeNode right;
   
    public TreeNode(int val) {
       this.val = val;
       this.left = this.right = null;
    }
}